selfservice/strategy/saml/handler.go

```
@@ -0,0 +1,374 @@
+		x.CSRFProvider
+		session.ManagementProvider
+		session.PersistenceProvider
+		errorx.ManagementProvider
+		config.Provider
+	}
+	HandlerProvider interface {
+		LogoutHandler() *Handler
+	}
+	Handler struct {
+		d  handlerDependencies
+		dx *decoderx.HTTP
+	}
+)
+
+type SessionData struct {
+	SessionID string
+}
+
+func NewHandler(d handlerDependencies) *Handler {
+	return &Handler{
+		d:  d,
+		dx: decoderx.NewHTTP(),
+	}
+}
+
+func (h *Handler) RegisterPublicRoutes(router *x.RouterPublic) {
+	h.d.CSRFHandler().IgnoreGlob(RouteBaseAcs + "/*")
+
+	router.GET(RouteMetadata, h.serveMetadata)
+	router.GET(RouteAuth, h.loginWithIdp)
+}
+
+// Handle /selfservice/methods/saml/metadata
+func (h *Handler) serveMetadata(w http.ResponseWriter, r *http.Request, ps httprouter.Params) {
+	config := h.d.Config()
+	pid := ps.ByName("provider")
+
+	if samlMiddlewares[pid] == nil {
+		if err := h.instantiateMiddleware(r.Context(), *config, pid); err != nil {
+			h.d.SelfServiceErrorManager().Forward(r.Context(), w, r, err)
+			http.Error(w, err.Error(), http.StatusInternalServerError)
+			return
+		}
+	}
+
+	buf, _ := xml.MarshalIndent(samlMiddlewares[pid].ServiceProvider.Metadata(), "", "  ")
+	w.Header().Set("Content-Type", "text/xml")
+	w.Write(buf)
+}
+
+// TODO Swagger
+func (h *Handler) loginWithIdp(w http.ResponseWriter, r *http.Request, ps httprouter.Params) {
+	// Middleware is a singleton so we have to verify that it exists
+	config := h.d.Config()
+	pid := ps.ByName("provider")
+
+	if samlMiddlewares[pid] == nil {
+		if err := h.instantiateMiddleware(r.Context(), *config, pid); err != nil {
+			h.d.SelfServiceErrorManager().Forward(r.Context(), w, r, err)
+			http.Error(w, err.Error(), http.StatusInternalServerError)
+			return
+		}
+	}
+
+	// We have to get the SessionID from the cookie to inject it into the context to ensure continuity
+	cookie, err := r.Cookie(continuity.CookieName)
+	if err != nil {
+		h.d.SelfServiceErrorManager().Forward(r.Context(), w, r, err)
```

Return missing, error will be ignored